MeshCoordMetadata [AVD-1629]

pp-mo opened the pull request:
## 🚀 Pull Request

### Description

Adds MeshCoordMetadata class (with tests)
Preliminary to creating a MeshCoord

Note: also fixed some problems I stumbled over in existing metadata testing.
Some of which  had already been copied into other tests! ...
There **_is_** a lot of boilerplate here : 
much of the  code in `lib/iris/tests/unit/experimental/ugrid/test_MeshCoordMetadata.py` is a total copy 
of that in `lib/iris/tests/unit/experimental/ugrid/test_ConnectivityMetadata.py` (i.e. ultimately from #3968)

As we are in a hurry, we are resolved to fix that "later"... 

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

stephenworsley reviewed (approved):
Looks good.

trexfeathers commented:
>Note: also fixed some problems I stumbled over in existing metadata testing.

Thanks for the catch @pp-mo!

pp-mo commented:
Sorry all -- I failed to include all the changes, at first.
Now fixed

stephenworsley commented on lib/iris/experimental/ugrid.py:
```diff
@@ -615,17 +615,130 @@ def equal(self, other, lenient=None):
+# of lenient metadata services.
+# TODO: when included in 'iris.common.metadata', install each one directly ?
+for cls in (ConnectivityMetadata, MeshCoordMetadata):
+    op_names_and_collections = {
```
It might make more sense to define `op_names_and_collections` outside the loop.

bjlittle replied:
@pp-mo If you do make convenience variables at the module level, could you also please purge them (`del`) from the namespace please 👍

pp-mo replied:
Awseome spot !
👍

pp-mo replied:
Just as an afterthought : I'm a bit weak on this...
Would you consider it equivalent to define an `__all__`, or use only underscore-prefixed things in the package namespace, or is it really that much better to totally tidy it ??

( I always felt that trying to fix this was somewhat swimming against the tide, given the way that imports mess it all up anyway )

pp-mo replied:
Done, I think.

bjlittle commented on lib/iris/experimental/ugrid.py:
```diff
@@ -615,17 +615,130 @@ def equal(self, other, lenient=None):
+            A list of different metadata member values.
+
+        """
+        # Perform "strict" difference for "cf_role", "start_index", "src_dim".
```
@pp-mo You just need to update this comment to align it with its members 👍

pp-mo replied:
👍 😊

pp-mo replied:
Done!

bjlittle commented:
@pp-mo Awesome stuff! A few suggested changes, but LGTM. Great work 😀

We can start using the `MeshCoordMetadata` in anger to see how it feels, and revisit `axis` at a later date (soonish) to see what we think about it... plus we may have some concrete use case data that helps sway up in an informed direction. Regardlessly, changing the makeup of the `MeshCoordMetadata` is a cheap and easy step to take now 👍

bjlittle reviewed (changes requested):
@pp-mo Awesome stuff! Just a few suggested changes, but LGTM. Great work 😀

We can start using the `MeshCoordMetadata` in anger to see how it feels, and revisit `axis` at a later date (soonish) to see what we think about it... plus we may have some actual concrete use case data that helps us weigh up and make an informed decision. 

Regardlessly, changing the makeup of the `MeshCoordMetadata` is a cheap and easy step to take now 👍

🎉 🥳🍻

stephenworsley reviewed (changes requested):
Just a couple more comments about the testing along with the existing comments. Otherwise, this looks good!

pp-mo commented:
Hoping that latest addresses @bjlittle comments. Ping!

Still thinking about @stephenworsley tests changes.
Will respond on that shortly..

pp-mo commented:
Rebased to fix problem with tests.

pp-mo commented on lib/iris/tests/unit/experimental/ugrid/test_MeshCoordMetadata.py:
```diff
@@ -0,0 +1,740 @@
+    def test_op_strict_different(self):
+        lmetadata = self.cls(**self.values)
+        right = self.values.copy()
+        right["long_name"] = self.dummy
```
Effectively I "inherited" all this from the test_ConnectivityMetadata.
I have now "rationalised" the use of the non-member metadata in these tests : uses of "units" and "var_name" have been replaced with "long_name".  I thought that made some sense, as long_name can be just anything whereas units=None would read back as "Unit('unknown')" if using a real object.

bjlittle commented on lib/iris/experimental/ugrid.py:
```diff
@@ -615,17 +615,130 @@ def equal(self, other, lenient=None):
-#: Convenience collection of lenient metadata difference services.
-SERVICES_DIFFERENCE.append(ConnectivityMetadata.difference)
-SERVICES.append(ConnectivityMetadata.difference)
+    _members = ("mesh", "location", "axis")
```
@pp-mo Whoa... are you really suggesting that `mesh` is metadata here?

This is the payload of the `MeshCoord`, right? If so, then I don't agree with this... it's akin to making `points` and `bounds` members of the `CoordMetadata` or `DimCoordMetadata` :-1: 

Or do you mean `mesh` to be something different?

trexfeathers replied:
Obviously having the actual `Mesh` as part of the metadata wouldn't work, but given the idea of shared `Mesh`es I'd say some sort of mesh identifier is an important component of the `MeshCoord` metadata.

bjlittle replied:
If `mesh` is a cheap comparison, then yes this could be part of the metadata, but until we know better I think we should err on the side of this not being part the metadata

bjlittle replied:
Also, `__eq__` hasn't been defined within `iris.experimental.ugrid.Mesh` for this very reason... we're kinda sitting on the fence until we know better

pp-mo replied:
Ok removed it

pp-mo commented:
Ok, now removed 'mesh' from the metadata, and rebased to accomodate the latest on the mesh-data-model target.
Also tidied up the 'service collections' adjustments, now all handled at the end of 'ugrid.py'

pp-mo commented:
@stephenworsley @bjlittle think we have addressed all previous issues now.
Sorry about the rebase.
Please re-review !

bjlittle commented:
@pp-mo and @stephenworsley LGTM :+1:
